node/impl/full/eth.go

```diff
@@ -811,12 +812,168 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
+// search till it gets within 100_000 gas
+func findGoodGas(
+	ctx context.Context,
+	smgr *stmgr.StateManager,
+	msgIn *types.Message,
+	priorMsgs []types.ChainMsg,
+	ts *types.TipSet,
+) (int64, error) {
+	msg := *msgIn
+
+	high := msg.GasLimit
+	low := msg.GasLimit
+
+	canSuccessed := func(limit int64) (bool, error) {
+		msg.GasLimit = limit
+
+		res, err := smgr.CallWithGas(ctx, &msg, priorMsgs, ts)
+		if err != nil {
+			return false, err
+		}
+
+		log.Infow("findGoodGas canSuccessed", "GasLimit", limit, "ExitCode", res.MsgRct.ExitCode)
+		if res.MsgRct.ExitCode == exitcode.SysErrorIllegalArgument {
+			log.Errorw("illegal argument")
+		}
+		if res.MsgRct.ExitCode == exitcode.Ok {
+			return true, nil
+		} else {
+			return false, nil
+		}
+	}
+
+	for {
+		ok, err := canSuccessed(high)
+		if err != nil {
+			return -1, err
+		}
+		if ok {
+			break
+		}
+
+		low = high
+		high = high * 2
+
+		if high > build.BlockGasLimit {
+			high = build.BlockGasLimit
+			break
+		}
+	}
+
+	log.Infow("findGoodGas narrow down to best GasLimit")
+	checkThreshold := (high / 100)
+	for (high - low) > checkThreshold {
+		median := (low + high) / 2
+		ok, err := canSuccessed(median)
+		if err != nil {
+			return -1, err
+		}
+
+		if ok {
+			high = median
+			log.Infow("findGoodGas Success", "GasLimit", median, "Diff", high-low)
+		} else {
+			low = median
+			log.Infow("findGoodGas Failure", "GasLimit", median, "Diff", high-low)
+		}
+
+		checkThreshold = (median / 100)
+	}
+
+	return high, nil
+}
+
+// runMessage is very similar to the `gasEstimateGasLimit` method which CallWithGas
+// except it has to handle errors a little different. When the message fails with
+// OutOfGas, error 33 (reverted), or Illegial Arugment we call findGoodGas
+func runMessage(
+	ctx context.Context,
+	cstore *store.ChainStore,
+	smgr *stmgr.StateManager,
+	mpool *messagepool.MessagePool,
+	msgIn *types.Message,
+	ts *types.TipSet,
+) (int64, error) {
+	msg := *msgIn
+	currTs := ts
+
+	log.Infow("runMessage starting", "GasLimit", msg.GasLimit)
+	// msg.GasLimit = build.BlockGasLimit
+	// msg.GasFeeCap = big.Zero()
+	// msg.GasPremium = big.Zero()
+
+	fromA, err := smgr.ResolveToDeterministicAddress(ctx, msgIn.From, currTs)
+	if err != nil {
+		return -1, xerrors.Errorf("getting key address: %w", err)
+	}
+
+	pending, ts := mpool.PendingFor(ctx, fromA)
+	priorMsgs := make([]types.ChainMsg, 0, len(pending))
+	for _, m := range pending {
+		if m.Message.Nonce == msg.Nonce {
+			break
+		}
+		priorMsgs = append(priorMsgs, m)
+	}
+
+	// Try calling until we find a height with no migration.
+	var res *api.InvocResult
+	for {
+		res, err = smgr.CallWithGas(ctx, &msg, priorMsgs, ts)
+		if err != stmgr.ErrExpensiveFork {
+			break
+		}
+		ts, err = cstore.GetTipSetFromKey(ctx, ts.Parents())
+		if err != nil {
+			return -1, xerrors.Errorf("getting parent tipset: %w", err)
+		}
+	}
+	if err != nil {
+		return -1, xerrors.Errorf("CallWithGas failed: %w", err)
+	}
```

Comment:
From 929 down to here is a copy of https://github.com/filecoin-project/lotus/blob/d6da796bf2d36560ad5b0a49a871d8e121ca8df4/node/impl/full/gas.go#L264-L292

I was thinking of pulling this out into its own function, but I would also need to return `priorMsgs` to be passed down further, as I need them when making additional calls to `CallWithSGas` inside of the currently named function `findGoodGas` which does the search.

So my time boxes refactor is to split the following into their own functions and use them here, and inside of `gasEstimateGasLimit`.

returns `priorMsgs`
https://github.com/filecoin-project/lotus/blob/d6da796bf2d36560ad5b0a49a871d8e121ca8df4/node/impl/full/gas.go#L269-L276

returns `res, err`
https://github.com/filecoin-project/lotus/blob/d6da796bf2d36560ad5b0a49a871d8e121ca8df4/node/impl/full/gas.go#L278-L292

Reply:
Actually there is probably not much of a reason to have two methods, and probably should just extract that entire block.